### ISSUE TITLE
support volatile queries without equality comparisons

### ISSUE THREAD

**Issue by nikomatsakis** [salsa-rs/salsa#56]
Currently, every "value" must implement `Eq`, so that we can compare if the value produced in revision R+1 differs from the value in revision R. For volatile queries, though, this isn't strictly needed -- it might be nice to offer a variant on volatile queries that does not require the value to implement `Eq`. Such a variant would always be assumed to change, even if it produces an "equal" value.

To implement this, we would move the "equality comparison" into the `MemoizationPolicy`, yielding something like this:

```rust
pub trait MemoizationPolicy<DB, Q>
where
    Q: QueryFunction<DB>,
    DB: Database,
{
    fn should_memoize_value(key: &Q::Key) -> bool;

    fn memoized_value_eq(old_value: &Q::Value, new_value: &Q::Value) -> bool;

    fn should_track_inputs(key: &Q::Key) -> bool;
}
```

Instead of doing `old_value == new_value`, we would invoke `memoized_value_eq`. Next, we remove the `Eq` (and `Hash` etc) bounds from the `Query` trait. Those bounds would rather move into the `MemoizationPolicy`.

Then we can offer a `untracked` memoization policy (which requires `Eq` and does comparisons) and a `volatile` (which is always assumed to change). Or something like that.

**Comment by nikomatsakis** [comment]
I created a [Zulip topic `#56 remove Eq bounds`][zulip] for discussing this.

[zulip]: https://salsa.zulipchat.com/#narrow/stream/146365-good-first-issue/topic/.2356.20remove.20.60Eq.60.20bounds

**Comment by matklad** [comment]
Another case I've hit when equality is not necessary is for "top-level" queries, which can't be a dependency of any other query. I don't think it's worth special-casing this case, it's just smth to keep in mind.

**Comment by nikomatsakis** [comment]
I *think* this issue can be marked closed. There were a few questions that came up in the implementation:

- Do we want a version of `input` that doesn't do "equality comparisons" (e.g., it just assumes that each call to `set` has a new value and updates the revision accodingly)? Maybe that should be the *only* kind of `input`? :)
- Similarly, do we want a version of "volatile" that *does* assume `Eq` (I was tentatively calling it "untracked")? You can always model this by wrapping a volatile with a memoized.

It *feels* like there is something a bit orthogonal here -- like there might be some orthogonal flag we can pull out about whether to compare values for equality. 

Since volatile is really just a memoized query with an untracked input, you could also imagine removing `volatile` as a thing and just having a method on `db` like `db.mark_volatile()` that a volatile query can invoke to force itself to be recomputed. 

At that point, memoized queries would just have some kind of flag about whether to compare their outputs for equality. If false, they don't require an `Eq` bound, but it means that if any inputs have changed, their output is *assumed* to change (since we can't test otherwise). You can also model *this*, of course, with a dummy wrapper type:

```rust
struct NoEq<T> {
    value: T
}

impl<T> PartialEq for NoEq<T> {
    fn partial_eq(&self, other: &NoEq<T>) -> bool { false }
}
```

but it feels a bit "odd" to do so.

**Comment by matklad** [comment]
>it seems like top-level queries could be just marked volatile 

You might still want to cache them, in theory. But yeah, for practical purposes I think volatile + frozen revision covers everything.

**Comment by nikomatsakis** [comment]
Yeah, that would correspond to the "untracked" variant I was thinking about

**Comment by nikomatsakis** [comment]
I'm contemplating just closing this. I think the real question is whether we want to expose 'more knobs' when it comes to the storage varieties.

**Comment by matklad** [comment]
@nikomatsakis should we use `Eq` trait at all?

I think we don't need strict equality, we need "maybe equal", and we need an ability to delegate that check to a hash function.

Hm, perhaps we need to bound values by `PartialEq`, and privides wrapper `NeverEq` (which return false) and `HashedEq` (which compare & store hashes (hm, or are hashes useless unless we serialzie values to disk))?

**Comment by nikomatsakis** [comment]
Idea: 

- Remove the "memoization policy" argument from `Derived` and instead have the *query type* implement that trait
- Export "memoization policy" as a non-plumbing thing users can access
- `storage memoized` etc just generate a suitable impl of the trait
- `storage custom` (or something) lets users implement it themselves

It then exposes the methods to be used for equality comparisons etc.

**Comment by matklad** [comment]
>Remove the "memoization policy" argument from Derived and instead have the query type implement that trait

I think we might need some kind of per-query execution memoization policy in the future.

For example in this code, I build a symbol index for all files in the project: https://github.com/rust-analyzer/rust-analyzer/blob/962a491829c2a803f5dc8c8c5e173baa89079e8b/crates/ra_analysis/src/imp.rs#L202-L210

Currently, I discard values of `syntax_tree` query after this operation, although, ideally, i'd love to avoid storing them in the frist place, to reduce peak memory usage. Note that this is not the same as making `SyntaxTree` query volatile: for documents which are curently open, I do want to preserve syntax trees.

Overall, I'd say we perhaps do the same thing we did with the input policy: use the simplest design that works for now?

**Comment by nikomatsakis** [comment]
@matklad 

> I think we might need some kind of per-query execution memoization policy in the future.

Agreed, though the design I proposed supports that...

> Currently, I discard values of syntax_tree query after this operation, although, ideally, i'd love to avoid storing them in the frist place, to reduce peak memory usage

Would the existing `storage dependencies` work for you then? If not, why not?

**Comment by matklad** [comment]
>Would the existing storage dependencies work for you then? If not, why not?

Not really: I want to memoize syntax trees, but only for *some* keys. The user typically edits a handful of files, but whole-project symbol search touches every file. So, for frequently accessed files I want to persist the syntax tree, but for "index the whole project once" scenario I don't want to remember the trees (except those that match the first scenario).

**Comment by Veykril** [comment]
This is no longer relevant in the current design
